libcaf_core/src/message.cpp

```diff
@@ -21,261 +21,164 @@
+  auto gmos = detail::global_meta_objects();
+  size_t data_size = 0;
+  for (size_t i = 0; i < ids_size; ++i) {
+    auto id = ids[i];
+    if (id >= gmos.size())
       return sec::unknown_type;
-    }
-    if (auto err = ptr->load(source))
+    auto& mo = gmos[id];
+    if (mo.type_name == nullptr)
+      return sec::unknown_type;
+    data_size += mo.padded_size;
+  }
+  auto vptr = malloc(sizeof(detail::message_data) + data_size);
+  auto ptr = new (vptr) detail::message_data(ids.move_to_list());
+  auto pos = ptr->storage();
+  auto types = ptr->types();
+  for (auto i = 0; i < ids_size; ++i) {
+    auto& meta = gmos[types[i]];
+    meta.default_construct(pos);
+    if (auto err = load(meta, source, pos)) {
+      auto rpos = pos;
+      for (auto j = i; j > 0; --j) {
+        auto& jmeta = gmos[types[j]];
+        jmeta.destroy(rpos);
+        rpos -= jmeta.padded_size;
+      }
+      ptr->~message_data();
+      free(vptr);
       return err;
-    dmd->append(std::move(ptr));
-    if (n != eos)
-      i = n + 1;
-    else
-      i = eos;
-  } while (i != eos);
-  if (auto err = source.end_object())
-    return err;
-  vals = detail::message_data::cow_ptr{std::move(dmd)};
-  return none;
+    }
+    pos += meta.padded_size;
+  }
+  data.reset(ptr, false);
+  return caf::none;
 }
 
 } // namespace
 
 error message::load(deserializer& source) {
-  return load_vals(source, vals_);
+  return load_data(source, data_);
 }
 
 error_code<sec> message::load(binary_deserializer& source) {
-  return load_vals(source, vals_);
+  return load_data(source, data_);
 }
 
 namespace {
 
 template <class Serializer>
 typename Serializer::result_type
-save_tuple(Serializer& sink, const type_erased_tuple& x) {
-  if (sink.context() == nullptr)
-    return sec::no_context;
-  // build type name
-  uint16_t zero = 0;
-  std::string tname = "@<>";
-  if (x.empty()) {
-    if (auto err = sink.begin_object(zero, tname))
-      return err;
-    return sink.end_object();
-  }
-  auto& types = sink.context()->system().types();
-  auto n = x.size();
-  for (size_t i = 0; i < n; ++i) {
-    auto rtti = x.type(i);
-    const auto& portable_name = types.portable_name(rtti);
-    if (portable_name == types.default_type_name()) {
-      std::cerr << "[ERROR]: cannot serialize message because a type was "
-                   "not added to the types list, typeid name: "
-                << (rtti.second != nullptr ? rtti.second->name()
-                                           : "-not-available-")
-                << std::endl;
-      return sec::unknown_type;
-    }
-    tname += '+';
-    tname += portable_name;
+save_data(Serializer& sink, const message::data_ptr& data) {
+  // Short-circuit empty tuples.
+  if (data == nullptr) {
+    uint16_t zero = 0;
+    return sink(zero);
   }
-  if (auto err = sink.begin_object(zero, tname))
+  // Write type information.
+  auto type_ids = data->types();
+  auto type_ids_size = static_cast<uint16_t>(type_ids.size());
```

Comment:
It is safe assume this fits because we limited `type_id`s to 16 bit and it uses it's value as an index?

Reply:
Yes, `type_ids.size()` returns the first element (16-bit integer) as a `size_t`. This cast simply converts back to the smaller type.